src/Umbraco.Core/Composing/ComponentCollection.cs

```diff
@@ -5,59 +5,61 @@
         {
-            foreach (IComponent component in this)
+            foreach (IAsyncComponent component in this)
             {
                 Type componentType = component.GetType();
-                using (!_profilingLogger.IsEnabled(Logging.LogLevel.Debug) ? null : _profilingLogger.DebugDuration<ComponentCollection>(
-                    $"Initializing {componentType.FullName}.",
-                    $"Initialized {componentType.FullName}.",
-                    thresholdMilliseconds: LogThresholdMilliseconds))
+
+                using (_profilingLogger.IsEnabled(Logging.LogLevel.Debug) is false
+                    ? null :
+                    _profilingLogger.DebugDuration<ComponentCollection>($"Initializing {componentType.FullName}.", $"Initialized {componentType.FullName}.", thresholdMilliseconds: LogThresholdMilliseconds))
                 {
-                    component.Initialize();
+                    await component.InitializeAsync(isRestarting, cancellationToken);
                 }
             }
         }
     }
 
-    public void Terminate()
+    public async Task TerminateAsync(bool isRestarting, CancellationToken cancellationToken)
     {
-        using (!_profilingLogger.IsEnabled(Logging.LogLevel.Debug) ? null : _profilingLogger.DebugDuration<ComponentCollection>(
-                   $"Terminating. (log components when >{LogThresholdMilliseconds}ms)", "Terminated."))
+        using (!_profilingLogger.IsEnabled(Logging.LogLevel.Debug)
+            ? null
+            : _profilingLogger.DebugDuration<ComponentCollection>($"Terminating. (log components when >{LogThresholdMilliseconds}ms)", "Terminated."))
         {
             // terminate components in reverse order
-            foreach (IComponent component in this.Reverse())
+            foreach (IAsyncComponent component in this.Reverse())
             {
                 Type componentType = component.GetType();
-                using (!_profilingLogger.IsEnabled(Logging.LogLevel.Debug) ? null : _profilingLogger.DebugDuration<ComponentCollection>(
-                    $"Terminating {componentType.FullName}.",
-                    $"Terminated {componentType.FullName}.",
-                    thresholdMilliseconds: LogThresholdMilliseconds))
+
+                using (_profilingLogger.IsEnabled(Logging.LogLevel.Debug) is false
+                    ? null
+                    : _profilingLogger.DebugDuration<ComponentCollection>($"Terminating {componentType.FullName}.", $"Terminated {componentType.FullName}.", thresholdMilliseconds: LogThresholdMilliseconds))
                 {
                     try
                     {
-                        component.Terminate();
-                        component.DisposeIfDisposable();
+                        await component.TerminateAsync(isRestarting, cancellationToken);
+                        await component.DisposeAsyncIfDisposable();
```

Comment:
Especially now that the `AsyncComponentBase` doesn't terminate on restarts (by default), this would cause the component to still get disposed 🤯 And if you really need to dispose resources on termination, you can already do so in the `TerminateAsync()` method anyway 🤔